geps/gep-1709.md

```diff
@@ -37,8 +45,304 @@ which implementations can prove they satisfy and be recognized for.
+Or the equivalent configuration using the Golang library:
+
+```go
+cSuite, err := suite.New(suite.Options{
+    GatewayClassName: "acme",
+    Profiles: sets.New(
+        Layer7,
+        Layer4,
+    ),
+    UnsupportedFeatures: sets.New(
+        suite.SupportHTTPResponseHeaderModification,
+        suite.SupportHTTPRouteMethodMatching,
+        suite.SupportHTTPRouteQueryParamMatching,
+    ),
+    // other options
+})
+require.NoError(t, err, "misconfigured conformance test suite")
+cSuite.Setup(t)
+
+for i := 0; i < len(tests.ConformanceTests); i++ {
+    test := tests.ConformanceTests[i]
+    test.Run(t, cSuite)
+}
+```
+
+> **NOTE**: `UnsupportedFeatures` must match support levels. Disabling features
+> which are in core will emit a warning and wont be reportable.
+
+> **NOTE**: In the future we may consider expanding the options to configure
+> and run the conformance test suite using a container image and/or via a
+> configuration file, however it was decided to hold off from doing any of
+> those in this iteration to avoid it getting to big (see the
+> [alternatives](#alternatives-considered) section for more thoughts).
+
+Some implementations may support more or less extended features than others,
+so in some cases it could be cumbersome to have to list ALL features that you
+_don't_ support so we optionally and inversely allow `SupportedFeatures` so
+you can pick which option makes sense to you, and under the hood the
+expressions will compile to the same overall list:
+
+```go
+cSuite, err := suite.New(suite.Options{
+    GatewayClassName: "acme",
+    Profiles: sets.New(
+        Layer7,
+        Layer4,
+    ),
+    SupportedFeatures: sets.New(
+        suite.SupportHTTPRouteMethodMatching,
+    ),
+    // other options
+})
+```
+
+> **NOTE**: The `UnsupportedFeatures` and `SupportedFeatures` fields are 
+> mutually exclusive.
+
+Once an implementation has integrated with the conformance test suite, they can
+move on to [certification](#certification) to report the results.
+
+[go-test]:https://go.dev/doc/tutorial/add-a-test
+[go]:https://go.dev
+[lib]:https://pkg.go.dev/sigs.k8s.io/gateway-api@v0.6.1/conformance/utils/suite
+
+### Certification
+
+Implementations will be able to report their conformance testing results using
+our [reporting process](#reporting-process). Implementations will be able to
+visibly demonstrate their conformance results on their downstream projects and
+repositories using our [certification process](#certification-process).
+
+#### Reporting Process
+
+When conformance tests complete a log message will be emitted:
+
+```console
+Success! Conformance profiles "Layer7" and "Layer4" passed for version v0.7.0.
+The following extended features were flagged as not supported for "Layer7Extended":
+- HTTPResponseHeaderModification
+- HTTPRouteQueryParamMatching
+```
+
+For the above output upstream Gateway API can manually generate or update a report
+in the following format:
+
+```json
+apiVersion: v1alpha1
```

Comment:
wdyt about each project/implementation's report file being a list of these sort of report objects? e.g.

```yaml
---
kind: ConformanceReport
apiVersion: v1alpha1
implementation:
  name: acmeorg-acme
  url: https://acmeorg.com
  version: v9.99.0
  contact:
  - @acme-org/maintainers
date: "2023-02-28 20:29:41+00:00"
gatewayAPIVersion: v0.6.1
profiles:
- name: layer4
  core:
    supportedFeatures:
    - ...
  extended:
    supportedFeatures:
    - ...
    unsupportedFeatures:
    - ...
- name: layer7
  core:
    supportedFeatures:
    - ...
  extended:
    supportedFeatures:
    - ...
    unsupportedFeatures:
    - ...
---
kind: ConformanceReport
apiVersion: v1alpha1
implementation:
  name: acmeorg-acme
  url: https://acmeorg.com
  version: v9.98.0
  contact:
  - @acme-org/maintainers
date: "2023-02-28 20:29:41+00:00"
gatewayAPIVersion: v0.6.0
profiles:
...
```

this would allow us to add the fields like date, version, etc. to each report as well since that detail is often relevant when assessing an implementation

Reply:
I can see advantages to this, particularly for the purpose of the human reporting via a PR part, and also making it really straightforward to build an archive several years down the road when these reports get significantly bigger. :thinking: 

Only thing I would say is that you've combined `supportedFeatures` and `unsupportedFeatures` in the same report for `extended`, but there is a note that had attempted to establish these as mutually exclusive earlier. We'll have to discuss more on the subject of how we represent extended features (and maybe even implementation-specific ones) but that doesn't need to hold this particular comment up.

@LiorLieberman and @arkodg had also expressed interest in contributing to this doc, so I would like to hear from there.

Reply:
ah yeah feel free to disregard that part on the reporting of what is actually passed/failed etc. that was just a quick idea without considering what may have already been updated here

but good to hear some of that metadata additions and the idea that we could generate a report per-run of the suite sounds useful 👍🏽

Reply:
I also considered suggesting something of the form of what sonobuoy or some other tool might output, but not sure if we want to go that far in reporting test result details etc., I don't remember exactly but I imagine using sonobuoy was discussed when starting the conformance framework

Reply:
I wouldn't be opposed to using that in time, if we found there were advantages. But at least for this very specific PR I'm not ready to completely switch things around yet. I would encourage you to make a PR after this one merges _just_ for pushing this change so that we can hyper-focus on it.

Reply:
my high level comment here is, we should separate out `what tests to run` with `what to report`.
* under `what tests to run` , we should continue to support the existing knobs of `SupportedFeatures`, `UnsuportedFeatures`, `SkipTests` and `SupportAllFeatures`. The user can either choose to use these historic knobs **OR** use the new `ConformanceProfile` knob. These flags should be validated.
* under `what to report`, it could be as granular as which tests passed or more coarse as which features passed (if all tests within that feature passed), but this logic is not tied to the input knobs, and is tied to - which tests ran, which passed, which failed. And the report is an explicit add on data output enabled via some `--report` like flag

Reply:
what you're suggesting seems to at least be close to where we're headed already @arkodg, do you feel there's anything specific in the doc that needs to be changed?

Reply:
yeah the doc says `Profiles` flag, in conjunction with `SkipTests` will result in an invalid report. Prefer if this tied to `what tests to run` instead, and said either use `Profile`  or use a mix of `SupportedFeatures, UnsuportedFeatures, SkipTests and SupportAllFeatures` to decide which tests to run. If not adhered, the tests wont start/or fail....

Reply:
^ can be added later as well as a follow up

Reply:
regardless of the exact reporting format of test results, seems useful to generally summarize and agree on somewhere in this doc what is useful for gw api maintainers/implementers/users to see in the conformance test report

my two cents:
- the project/version/test run metadata above to identify a particular test result ^
- conformance profile test run results overall summary for each "profile"
  - implementation is deemed fully conformant with the profile if the tests all passed 
  - partial if some passed/some were skipped (IMO this is useful to know for esp. the extended profile)
  - no conformance if none passed or none were run (i.e. an L4 only controller might skip all L7 tests)
  - include summary stats, e.g. total number of tests, how many passed, skipped, failed
- conformance profile test run detailed results (in case of partial conformance etc.)
  - for each test, test name, short description like we have [here](https://github.com/kubernetes-sigs/gateway-api/blob/69fabd994640debc2d9c00d4a086cdee0bbe9a22/conformance/tests/gateway-invalid-route-kind.go#L35-L36)
  - did it pass, fail, or was it skipped
  - can include test filename for people to do closer inspection if needed
  - can add detailed failure messages, etc. if we want too

Reply:
(IMO seems ok to skip tests, that shouldn't necessarily make the output "invalid" but will mean you end up with a report that your implementation is not fully conformant)

Reply:
I'm on board with the notion of separating what to test from what to report conceptually as @arkodg mentions above, we could e.g. start with no changes to the test suite invocation and iteratively add convenience flags to let you run subsets of tests that correspond to "profiles"

Reply:
After thinking it over I was persuaded by the list of separate `ConformanceReports` idea, and did a reworking of the `ConformanceReport` API which includes that and somewhat follows the bullet points you listed above, @sunjayBhatia.

My hope is that this will be satisfactory for all stakeholders. I'm going to consider this comment thread resolved, as it's gotten quite long and hard to follow, but if any of you _don't_ find the new version satisfactory please create some new comments **ideally with explicitly and prescriptive suggestions about what specifically you would like to see changed** as that can be incredibly helpful to keeping up velocity on these GEPs. Also keep in mind that we're still in a `provisional` state so nothing that merges here is going to be implemented yet.